machxo2/pack.cc

```diff
@@ -19,326 +18,1567 @@
+                    int distance = std::abs(drvloc.x - pllloc.x) + std::abs(drvloc.y - pllloc.y);
+                    if (distance < closest_distance) {
+                        closest_pll = bel;
+                        closest_distance = distance;
+                    }
+                }
+                if (closest_pll == BelId())
+                    log_error("failed to place PLL '%s'\n", ci->name.c_str(ctx));
+                available_plls.erase(closest_pll);
+                ci->attrs[id_BEL] = ctx->getBelName(closest_pll).str(ctx);
+            }
+        }
+        // Place PLLs driven by logic, etc, randomly
+        for (auto &cell : ctx->cells) {
+            CellInfo *ci = cell.second.get();
+            if (ci->type == id_EHXPLLJ && !ci->attrs.count(id_BEL)) {
+                if (available_plls.empty())
+                    log_error("failed to place PLL '%s'\n", ci->name.c_str(ctx));
+                BelId next_pll = *(available_plls.begin());
+                available_plls.erase(next_pll);
+                ci->attrs[id_BEL] = ctx->getBelName(next_pll).str(ctx);
+            }
         }
     }
 
-    ctx->cells[const_cell->name] = std::move(const_cell);
+    // Check if two nets have identical constant drivers
+    bool equal_constant(NetInfo *a, NetInfo *b)
+    {
+        if (a == nullptr && b == nullptr)
+            return true;
+        if ((a == nullptr) != (b == nullptr))
+            return false;
+        if (a->driver.cell == nullptr || b->driver.cell == nullptr)
+            return (a->driver.cell == nullptr && b->driver.cell == nullptr);
+        if (a->driver.cell->type != id_GND && a->driver.cell->type != id_VCC)
+            return false;
+        return a->driver.cell->type == b->driver.cell->type;
+    }
+
+    struct EdgeClockInfo
+    {
+        CellInfo *buffer = nullptr;
+        NetInfo *unbuf = nullptr;
+        NetInfo *buf = nullptr;
+    };
+
+    std::map<std::pair<int, int>, EdgeClockInfo> eclks;
+    std::map<NetInfo *, int> bridge_side_hint;
+
+    void tie_zero(CellInfo *ci, IdString port)
+    {
 
-    for (auto dn : dead_nets) {
-        ctx->nets.erase(dn);
+        if (!ci->ports.count(port)) {
+            ci->ports[port].name = port;
+            ci->ports[port].type = PORT_IN;
+        }
+        IdString name = ctx->id(ci->name.str(ctx) + "$zero$" + port.str(ctx));
+
+        auto zero_cell = std::make_unique<CellInfo>(ctx, name, id_GND);
+        NetInfo *zero_net = ctx->createNet(name);
+        zero_cell->addOutput(id_GND);
+        zero_cell->connectPort(id_GND, zero_net);
+        ci->connectPort(port, zero_net);
+        new_cells.push_back(std::move(zero_cell));
     }
-}
 
-static bool is_nextpnr_iob(Context *ctx, CellInfo *cell)
-{
-    return cell->type == ctx->id("$nextpnr_ibuf") || cell->type == ctx->id("$nextpnr_obuf") ||
-           cell->type == ctx->id("$nextpnr_iobuf");
-}
+    int lookup_delay(const std::string &del_mode)
```

Comment:
are you actually supporting the IO DELAYs? even if you are, have you checked these values actually make sense for machxo2?